server/handlers/project_command_output_handler.go

```diff
@@ -141,20 +130,22 @@ func (p *AsyncProjectCommandOutputHandler) clearLogLines(pull string) {
 }
 
 func (p *AsyncProjectCommandOutputHandler) addChan(ch chan string, pull string) {
-	p.receiverBuffersLock.Lock()
-	if p.receiverBuffers[pull] == nil {
-		p.receiverBuffers[pull] = map[chan string]bool{}
-	}
-	p.receiverBuffers[pull][ch] = true
-	p.receiverBuffersLock.Unlock()
-
 	p.projectOutputBuffersLock.RLock()
 	buffer := p.projectOutputBuffers[pull]
 	p.projectOutputBuffersLock.RUnlock()
 
 	for _, line := range buffer {
```

Comment:
I'm curious how the buffer would behave if there's new messages coming in to the `projectOutputBuffers` while we send messages into the `ch` channel.  Since we only add the channel `ch` to the `receiverBuffers[pull]` later in the method, the writelogLine method would not forward the new messages into the channel neither will we have it copied over from the output buffer unless the buffer is being passed by reference.  If it is being passed by reference should we just leave it inside the Rlock to ensure two goroutines don't read and write in the buffer at the same time? https://github.com/lyft/atlantis/blob/60e3d5947a37203e77b97845fff363a9cf4a4d9e/server/handlers/project_command_output_handler.go#L154-L166

Reply:
it's not a reference, the underlying array gets copied.  Realistically we should be using `sync.Map` here, not sure why we aren't.  I plan on moving to that in a future PR.

The problem you're mentioning would happen regardless of whether it's in the Rlock since there is a chance we don't get the receiver buffer lock in this method and then lines are stored in the buffer but not forwarded to the channel.  Something we'll have to fix with a future refactor to this. These two data objects should be accessed atomically imo.

Reply:
Gotcha! Thanks for the clarification :)